<!DOCTYPE html><html lang="en"><head><script async src="https://www.googletagmanager.com/gtag/js?id=UA-114897551-4"></script><script>window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'UA-114897551-4');
</script><script type="text/javascript" src="//platform-api.sharethis.com/js/sharethis.js#property=5ac2443d1fff98001395ab6c&amp;product=sticky-share-buttons" async="async"></script><title>Error Handling - Intermediate Python Programming p.23 | Coder Coacher - Coaching Coders</title><meta content="Error Handling - Intermediate Python Programming p.23 - All technical stuff in one place" name="description"><meta name="keywords" content="education, coding, programming, technology, nodejs, mongodb, software, computer science, engineering, teaching, coaching, coder, learning, java, kotlin"><meta name="viewport" content="width=device-width, initial-scale=1.0"><link rel="stylesheet" href="/css/font.css"><link rel="stylesheet" href="/css/bootstrap.css"><link rel="stylesheet" href="/css/style.css"><link rel="stylesheet" href="/css/coder-coacher.css"></head><body><div class="container-fluid"><h1 class="site-title"><a href="/">Coder Coacher</a></h1><hr><h4 class="site-subtitle text-right">Coaching Coders</h4></div><div id="amzn-assoc-ad-99d6751e-2392-4004-ad16-73aa8385d9d0"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=99d6751e-2392-4004-ad16-73aa8385d9d0"></script><div class="post__breadcrumb"><div class="container"><ol class="breadcrumb"><li><a href="/">Coder Coacher</a></li><li><a href="/sentdex/">sentdex</a></li><li class="active">â¤µ</li></ol></div></div><h2 class="post__title"><b>Error Handling - Intermediate Python Programming p.23</b></h2><h5 class="post__date">2016-12-12</h5><div class="container"><div class="video-responsive"><iframe width="560" height="315" src="https://www.youtube.com/embed/7-2rNHi9EfU" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe></div><div class="post__text">what is going on everybody welcome to
part 23 of the intermediate python
tutorial series in this part what we're
going to be talking about is handling
errors and actually getting all the
information that we might need
especially when we are doing something
like logging as we covered in the
previous tutorial so to start you know
typically let's save a try and accept it
and in the try already a plus B and then
in the exception as e let's just print
string mmm okay save and run that the
the error that we get is just name a is
not defined now that can be sort of
useful but again in a larger program
that's them a could be all over the
place we might not have you know we're
not getting we don't know where in the
program was this it and so that that can
be problematic
so especially as opposed to if we well
we can actually leave that exception
here so we'll hit that exception and
then we'll hit another exception right
in here we actually get a trace back and
we're given all kinds of lovely
information where did it happen
what line happened what was the code
basically and what was the error but
when we just return string e all we get
his name is not a find
so I'm gonna close this and and
basically what we're going to solve here
is it is this problem so first we're
going to do is we're going to import sis
now what's this is gonna let us do is
actually access the trace back itself
the actual exception error and then we
can and then we cannot start returning
information from that error
so rather than print string e we can do
something like this like let's print sis
dot exe for exception info and let's
just do that for now so have a run that
and as you can see it's actually a tuple
of some information for us we've got
what kind of error was it and then we
actually literally get the error and
then we get this trace back object blah
blah the trace back object is the actual
information of like the line number and
other bits of information for us let's
close that and instead what we're gonna
do is we're gonna slice this up now
interestingly enough I read from the
Internet's that in in the exception it's
unwise to save the slices to variables
so for example you might you might
wanted well first let's just do this let
me just print it out so well let's just
print the slices so zero will do it will
print that and now I'm going to paste
this three times or two times rather one
and two so the zeroth was the error
this was the name error and then this
was this would be like that that trace
back information and we can say TB
underscore line number line no for line
number
let's save and run that okay you get
name err name a is not find line four
now as I started to say before I want to
just show you real quick it's unwise to
save these values to variables because
if you have an exception in the
exception itself you would have a
circular reference to the variables and
they would never get garbage collected
so you really want to just slice the
tuple like this to not save it to a
variable interesting information anyway
and that's fine we can leave it like
that so so this is our new exception and
then instead maybe what we would say is
something like there's a print error :
and then we're gonna do well I'm
formatting here or string formatting
rather and then we'll say online
something dot format and then we'll
format just all these this comma I'm
just gonna cut enter down there
copy paste comma and then this one here
paste I'm gonna delete what we had there
before and now you get error you know
name err name is not defined line four
okay a lot more useful of an error in
our program here it's very simple
and when we would have known immediately
where a was because a is only referenced
one time in this entire script but if
you have a much larger program you might
have a reference in hundreds of places
where where are you talking about
so very cool now probably you know like
I said we're most likely to be using
this in like logging logging situation
so we can import logging here and then
let's create let's define error handling
and then we can take this basically
everything that was in that print
statement and instead let's just return
that information and then we could say
logging dot error and then we just do
error underscore so cool let's save that
run it and there you have it so now we
have not the entire trace back which you
know could be pages long and it we don't
want all of that saved necessarily to a
file now we've just got error this is
kind of redundant we probably could do
we could probably do like this vlogging
already gave us that error you know the
level basically so err the route
what kind of Nate what kind of error is
it name is not a fine line 11 much more
useful obviously this one we just kind
of put out to the console but of course
with logging you could say you it's
gonna be saved to a file or something
like that
so yeah I just wanted to show that
really quickly because I think it's a
lot more useful especially when you when
using logging and stuff like that so if
you have any questions comments concerns
suggestions whatever feel free to leave
them below otherwise I will see you in
another tutorial</div></div><div class="container-fluid bottom-ad"><div id="amzn-assoc-ad-6a809dda-347a-4187-8a86-91faf94575da"></div><script async src="//z-na.amazon-adsystem.com/widgets/onejs?MarketPlace=US&amp;adInstanceId=6a809dda-347a-4187-8a86-91faf94575da"></script></div><div class="text-center">We are a participant in the Amazon Services LLC Associates Program, an affiliate advertising program designed to provide a means for us to earn fees by linking to Amazon.com and affiliated sites.</div><script>(function(w, d){
    var b = d.getElementsByTagName('body')[0];
    var s = d.createElement("script"); s.async = true;
    var v = !("IntersectionObserver" in w) ? "8.6.0" : "10.4.2";
    s.src = "https://cdnjs.cloudflare.com/ajax/libs/vanilla-lazyload/" + v + "/lazyload.min.js";
    w.lazyLoadOptions = {};
    b.appendChild(s);
}(window, document));</script></body></html>